charmhelpers/contrib/openstack/utils.py

```diff
@@ -1599,6 +1599,14 @@ def pause_unit(assess_status_func, services=None, ports=None,
         services=services,
         charm_func=charm_func)
     set_unit_paused()
+
+    # construct a check function, making it retry with a delay if ports are open
+    @decorators.retry_on_predicate(5, bool, base_delay=0.1)
+    def retry_port_check(ports):
+        _, ports_states = _check_listening_on_ports_list(ports)
+        return not any(ports_states)
+    if ports:
+        retry_port_check(ports)  # block until check is successful or times out
```

Comment:
I'm confused; it looks like ports wasn't used before (but still existed) and now does.  So AFAICT it waits for a bit checking the port, and if it's still not closed then it just moves on?  No exception is raised?  This function is now very confusing.  Shouldn't "manage_payload_services" actually do the job of checking that the ports are close ... i.e. it's supposed to manage the payload service?  I think this is blurring what this function is supposed to do?